R/classes-model.R

```diff
@@ -310,23 +310,80 @@ predict.Model <- function(object, mfi, ...) {
 #'   If `TRUE`, the MFI range is calculated from all analytes; if `FALSE`, the MFI range is calculated only for the current analyte
 #'   Defaults to `FALSE`
+#' @param detect_high_dose_hook (`logical(1)`) If `TRUE`, the high dose hook effect is detected and handled.
+#' For more information, please see the \link[PvSTATEM]{handle_high_dose_hook} function documentation.
 #' @param ... Additional arguments passed to the model
 #'
 #' @return (`Model()`) Standard Curve model
 #'
 #' @export
-create_standard_curve_model_analyte <- function(plate, analyte_name, data_type = "Median", source_mfi_range_from_all_analytes = FALSE, ...) {
+create_standard_curve_model_analyte <- function(plate, analyte_name,
+                                                data_type = "Median",
+                                                source_mfi_range_from_all_analytes = FALSE,
+                                                detect_high_dose_hook = TRUE,
+                                                ...) {
   mfi <- plate$get_data(analyte_name, "STANDARD CURVE", data_type = data_type)
   dilutions_numeric <- plate$get_dilution_values("STANDARD CURVE")
 
-  mfi_source <- ifelse(source_mfi_range_from_all_analytes, "ALL", analyte_name)
+  if (detect_high_dose_hook) {
+    sample_selector <- handle_high_dose_hook(mfi, dilutions_numeric)
+    mfi <- mfi[sample_selector]
+    dilutions_numeric <- dilutions_numeric[sample_selector]
+  }
 
+  mfi_source <- ifelse(source_mfi_range_from_all_analytes, "ALL", analyte_name)
   mfi_min <- min(plate$get_data(mfi_source, "STANDARD CURVE", data_type = data_type), na.rm = TRUE)
   mfi_max <- max(plate$get_data(mfi_source, "STANDARD CURVE", data_type = data_type), na.rm = TRUE)
 
   Model$new(analyte_name, dilutions_numeric, mfi, mfi_min = mfi_min, mfi_max = mfi_max, ...)
 }
+
+
+#' @title Detect and handle the high dose hook effect
+#'
+#' @description
+#' Typically, the MFI values associated with standard curve
+#' samples should decrease as we dilute the samples. However,
+#' sometimes in high dilutions, the MFI presents a non monotonic behavior.
+#' In that case, MFI values associated with dilutions above (or equal to)
+#' `high_dose_threshold` should be removed from the analysis.
+#'
+#' For the `nplr` model the recommended number of standard curve samples
+#' is at least 4. If the high dose hook effect is detected but the number
+#' of samples below the `high_dose_threshold` is lower than 4,
+#' additional warning is printed and the samples are not removed.
+#'
+#' The function returns a logical vector that can be used to subset the MFI values.
+#'
+#' @param mfi (`numeric()`)
+#' @param dilutions (`numeric()`)
+#' @param high_dose_threshold (`numeric(1)`) MFI values associated
+#' with dilutions above this threshold should be checked for the high dose hook effect
+#'
+#' @return sample selector (`logical()`)
+handle_high_dose_hook <- function(mfi, dilutions, high_dose_threshold = 1 / 200) {
```

Comment:
how was the 1/200, choosen as deafault? I think we can discuss it tomorrow

Reply:
In the email conversation with Micheal
```
Can you please include a routine that tests for increasing MFI in the first dilutions (1/50, 1/100, 1/200).
```